### PR TITLE
2818 Update Indicator - 2

### LINKED ISSUE #2818: As tech lead, I need TDP to reject files that do not have an update indicator == D
**Description:**

Currently, the legacy system (TDRS) handles re-submitted records based on the [update indicator](https://www.acf.hhs.gov/sites/default/files/documents/ofa/transmission_file_header_trailer_record.pdf) in the header record. The rules are as follows:

- indicator = N for new records to be added to the db for a given fiscal period, program type, and data type
  - STTs may use this indicator as a mechanism for submit a subset of records
- indicator = U for records  that are being re-submitted to update existing records in the db for a given fiscal period, program type, and data type 
  - STTs often use this indicator as a mechanism for submit a subset of records
- indicator = D for records that are being re-submitted to delete and overwrite any existing records in the db for a given fiscal period, program type, and data type

TDP does not currently consider the update indicator in the pre-parsing logic; additionally we have been advising the STTs in our knowledge center to always use the D indicator for re-submissions to help them start to practice sending all records that they intend to submit for federal reporting. 

We now need a mechanism to ensure TDP is not parsing and storing data from "incomplete" files in the db. 

**Acceptance Criteria:** 
_Create a list of functional outcomes that must be achieved to complete this issue_

- [ ] If the update indicator != D, reject and do not parse the file. 
- [ ] feedback report should indicate this as a cat 1 error. 
- [x] OFA updates the header record documentation to remove the N and U options as valid possible values
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

**Tasks:**
_Create a list of granular, specific work items that must be completed to deliver the desired outcomes of this issue_ 

- [ ] _Task 1_
- [ ] _Task 2_
- [ ] _Task 3_
- [ ] Run Testing Checklist and confirm all tests pass

**Notes:**
_Add additional useful information, such as related issues and functionality that isn't covered by this specific issue, and other considerations that will be helpful for anyone reading this_ 

- _Note 1_
- _Note 2_
- _Note 3_

**Supporting Documentation:**
_Please include any relevant log snippets/files/screen shots_

- _Doc 1_
- _Doc 2_

**Open Questions:**
_Please include any questions or decisions that must be made before beginning work or to confidently call this issue complete_ 

- A lower lift would be to remove the N and U as valid possible options  and keep parsing, so STTS have a chance to correct it? 
- _Open Question 2_

### PR DESCRIPTION
**Pull request by atrimpe**
## Summary of Changes
Altered header to only allow D, updated test data files, added test
Pull request closes #2818 

## How to Test
_List the steps to test the PR_
These steps are generic, please adjust as necessary.
```
cd tdrs-frontend && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d
cd tdrs-backend && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d 
```

1. Open http://localhost:3000/ and sign in.
1. Proceed with functional tests as described herein.
1. Test steps should be captured in the demo GIF(s) and/or screenshots below.
> *Demo GIF(s) and screenshots for testing procedure*

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] If the update indicator != D, reject and do not parse the file. 
- [x] feedback report should indicate this as a cat 1 error. 
_it shows as a cat 2 error but functions as a cat 1 error insofar as the data is not parsed_
+ [x] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

**Review by elipe17 — APPROVED** [review]

**Review by jtimpe — APPROVED** [review]

**Comment by atrimpe** [comment]
@ADPennington all good to go

**Comment by ADPennington** [comment]
> @ADPennington all good to go

apologies for the delay in review. some conflicts here @atrimpe. I can review as soon as these are resolved.

**Comment by atrimpe** [comment]
> > @ADPennington all good to go
> 
> apologies for the delay in review. some conflicts here @atrimpe. I can review as soon as these are resolved.

@ADPennington all good

**Review by ADPennington — COMMENTED** [review]
@atrimpe this is in good shape in terms of rejecting and not parsing files with the incorrect update indicator, but a couple of nits below. 

- the error seems to be raised as a category 2 error instead of a category 1 error (which is one of the ACs). It appears that this is because of how the code is organized, where most of the elements in the header records are checked as `postparsing_validators`. If true, I think this is okay for now. 
- the error message on this could be better. It is not clear what the problem is (e.g. `N does not match D.`). @reitermb what do you think? If we do not make a custom validator for this field, we'll need to be prepared with guidance in the TDP Knowledge Center, because this is a big change for STTs, one that will likely yield questions. 


testing notes [here](https://hackmd.io/Ban2OmoQT_unHqLLLngUpg?view#2818-Update-Indicator---2)

qasp approval on hold until @reitermb weighs in.

**Comment by reitermb** [comment]
> @atrimpe this is in good shape in terms of rejecting and not parsing files with the incorrect update indicator, but a couple of nits below.
> 
> * the error seems to be raised as a category 2 error instead of a category 1 error (which is one of the ACs). It appears that this is because of how the code is organized, where most of the elements in the header records are checked as `postparsing_validators`. If true, I think this is okay for now.
> * the error message on this could be better. It is not clear what the problem is (e.g. `N does not match D.`). @reitermb what do you think? If we do not make a custom validator for this field, we'll need to be prepared with guidance in the TDP Knowledge Center, because this is a big change for STTs, one that will likely yield questions.
> 
> testing notes [here](https://hackmd.io/Ban2OmoQT_unHqLLLngUpg?view#2818-Update-Indicator---2)
> 
> qasp approval on hold until @reitermb weighs in.

I think one light tweak we could add here is ```Header Update Indicator: [x] does not match D``` so we're providing a "codeword" of sorts classifying what this is, nodding to layout docs, terminology in the knowledge center etc... 

Re: The Knowledge Center I do think it's worth reframing [this page](https://reitermb.github.io/Test/knowledge-center/exporting-complete-data-using-ftanf.html) around Submission Requirements more broadly—or adding a submission requirements page that will further link to that existing guide. 

#2847 could also link to guidance around this in the section that breaks down what the header/trailer are.

Thoughts on this @ADPennington ?

**Review by raftmsohani — APPROVED** [review]
LGTM!

**Comment by ADPennington** [comment]
> > @atrimpe this is in good shape in terms of rejecting and not parsing files with the incorrect update indicator, but a couple of nits below.
> > 
> > * the error seems to be raised as a category 2 error instead of a category 1 error (which is one of the ACs). It appears that this is because of how the code is organized, where most of the elements in the header records are checked as `postparsing_validators`. If true, I think this is okay for now.
> > * the error message on this could be better. It is not clear what the problem is (e.g. `N does not match D.`). @reitermb what do you think? If we do not make a custom validator for this field, we'll need to be prepared with guidance in the TDP Knowledge Center, because this is a big change for STTs, one that will likely yield questions.
> > 
> > testing notes [here](https://hackmd.io/Ban2OmoQT_unHqLLLngUpg?view#2818-Update-Indicator---2)
> > qasp approval on hold until @reitermb weighs in.
> 
> I think one light tweak we could add here is `Header Update Indicator: [x] does not match D` so we're providing a "codeword" of sorts classifying what this is, nodding to layout docs, terminology in the knowledge center etc...
> 
> Re: The Knowledge Center I do think it's worth reframing [this page](https://reitermb.github.io/Test/knowledge-center/exporting-complete-data-using-ftanf.html) around Submission Requirements more broadly—or adding a submission requirements page that will further link to that existing guide.
> 
> #2847 could also link to guidance around this in the section that breaks down what the header/trailer are.
> 
> Thoughts on this @ADPennington ?

thanks @reitermb -- i think this would require a custom validator. we're currently using the `matches` validator which is used in several other places. @atrimpe @elipe17  is it feasible to make a custom validator for use here, per Miles' suggestion? 

Also @reitermb i agree that we can plan for #2847 to handle guidance around this. I suspect that this PR will in the release for next week, will 2847 be ready by then? or will we have a draft hackmd ready?

**Comment by ADPennington** [comment]
currently blocked pending feedback from dev team. @atrimpe @elipe17

**Comment by reitermb** [comment]
> Also @reitermb i agree that we can plan for #2847 to handle guidance around this. I suspect that this PR will in the release for next week, will 2847 be ready by then? or will we have a draft hackmd ready? 

My goal is to have the content hack.md in review early next week before I'm partially OOO later that week/the next. My stretch goal will include the PR as well but depends on some in flight roadmap/backlog work with Rob & Andrew and tying up the submission history content PR (I'm shooting to have requested revisions on that in tomorrow).

**Comment by elipe17** [comment]
> > > @atrimpe this is in good shape in terms of rejecting and not parsing files with the incorrect update indicator, but a couple of nits below.
> > > 
> > > * the error seems to be raised as a category 2 error instead of a category 1 error (which is one of the ACs). It appears that this is because of how the code is organized, where most of the elements in the header records are checked as `postparsing_validators`. If true, I think this is okay for now.
> > > * the error message on this could be better. It is not clear what the problem is (e.g. `N does not match D.`). @reitermb what do you think? If we do not make a custom validator for this field, we'll need to be prepared with guidance in the TDP Knowledge Center, because this is a big change for STTs, one that will likely yield questions.
> > > 
> > > testing notes [here](https://hackmd.io/Ban2OmoQT_unHqLLLngUpg?view#2818-Update-Indicator---2)
> > > qasp approval on hold until @reitermb weighs in.
> > 
> > 
> > I think one light tweak we could add here is `Header Update Indicator: [x] does not match D` so we're providing a "codeword" of sorts classifying what this is, nodding to layout docs, terminology in the knowledge center etc...
> > Re: The Knowledge Center I do think it's worth reframing [this page](https://reitermb.github.io/Test/knowledge-center/exporting-complete-data-using-ftanf.html) around Submission Requirements more broadly—or adding a submission requirements page that will further link to that existing guide.
> > #2847 could also link to guidance around this in the section that breaks down what the header/trailer are.
> > Thoughts on this @ADPennington ?
> 
> thanks @reitermb -- i think this would require a custom validator. we're currently using the `matches` validator which is used in several other places. @atrimpe @elipe17 is it feasible to make a custom validator for use here, per Miles' suggestion?
> 
> Also @reitermb i agree that we can plan for #2847 to handle guidance around this. I suspect that this PR will in the release for next week, will 2847 be ready by then? or will we have a draft hackmd ready?

@ADPennington, @reitermb, have a look at what I did wrt the `matches` validator. Since the cat1 PR merged I was able to leverage that to get the friendly name into the error message. I kept the friendly name generic for the moment since [2801](https://github.com/raft-tech/TANF-app/issues/2801) has a flushed out update to all friendly names.

**Comment by ADPennington** [comment]
@elipe17 @raftmsohani  terraform pipeline issue during deployment.

File: tdrs-backend/tdpservice/parsers/schema_defs/header.py [hunk 1]
```diff
@@ -117,12 +117,12 @@
         Field(
             item="10",
             name="update",
-            friendly_name="update",
+            friendly_name="update indicator",
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/schema_defs/header.py:120]: @elipe17 thank you for this! `HEADER update indicator: N does not match D.`

**Review by ADPennington — APPROVED** [review]
@atrimpe @elipe17 @reitermb this is approved and ready to merge. 

one note: the AC on this ticket was to raise a cat 1 (pre-parsing error) when update indicator isnt D. technically, this raised a cat 2 error, which i think is okay because: (1) error types will be removed from the error feedback report as part of #2897 and (2) the expected behavior of a cat 1 error (which is to not process files with incorrect header) is implemented . 🚀 


test notes [here](https://hackmd.io/Ban2OmoQT_unHqLLLngUpg?view#2818-Update-Indicator---2)

**Pull request merged**